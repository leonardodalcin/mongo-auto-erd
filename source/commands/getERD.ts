import { makeEntity } from '@entity/makeEntity'
import { IEntity } from '@interfaces/IEntity'
import { getDB } from '@mongo/getDB'
import { getDBCollectionNames } from '@mongo/getDBCollectionNames'
import { Spinner } from 'cli-spinner'
import { mapReduceCollectionProperties } from 'source/mongo/mapReduceCollectionProperties'

const spinner = new Spinner()
spinner.setSpinnerString(14)

export async function getERD(
  mongoURI: string,
  databaseName: string,
  outdir: string
): Promise<IEntity[]> {
  spinner.start()
  await getDB(mongoURI, databaseName)
  const collectionNames = await getDBCollectionNames()
  const entities = []
  for (const collectionName of collectionNames) {
    entities.push(await makeEntity(collectionName))
  }
  spinner.stop()
  return entities
}
