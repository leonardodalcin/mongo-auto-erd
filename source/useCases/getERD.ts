import { IEntity } from '@interfaces/IEntity'
import { getDB } from '@mongoWrapper/getDB'
import { getDBCollectionNames } from '@mongoWrapper/getDBCollectionNames'
import { getRelationshipTargetCollectionName } from '@mongoWrapper/getRelationshipTargetCollectionName'
import { getRelationshipType } from '@mongoWrapper/getRelationshipType'
import { mapReduceCollectionProperties } from '@mongoWrapper/mapReduceCollectionProperties'
import { Spinner } from 'cli-spinner'

const spinner = new Spinner()
spinner.setSpinnerString(14)

export async function getERD(
  mongoURI: string,
  databaseName: string,
  outdir: string
): Promise<void> {
  await getDB(mongoURI, databaseName)
  spinner.start()
  const collectionNames = await getDBCollectionNames()
  for (const collectionName of collectionNames) {
    const entity: IEntity = {
      collectionName,
      properties: [],
      relationships: []
    }

    const mapReducedProperties = await mapReduceCollectionProperties(
      collectionName
    )
    for (const property of mapReducedProperties) {
      if (property.values.some((value) => value.type === 'objectId')) {
        entity.relationships.push({
          propertyName: property.name,
          sourceCollectionName: collectionName,
          targetCollectionName:
            (await getRelationshipTargetCollectionName(property)) || '',
          type: (await getRelationshipType(property)) || ''
        })
      }
    }
  }
  spinner.stop()
}
