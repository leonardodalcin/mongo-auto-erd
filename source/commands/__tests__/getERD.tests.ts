import { getERD } from '@commands/getERD'
import { populateDatabase } from '@tests/dbSetup'

describe('getERD', () => {
  beforeAll(async () => {
    await populateDatabase()
  })
  it('should match snapshot', async () => {
    const erd = await getERD(
      await global.mongod.getUri(),
      await global.mongod.getDbName()
    )
    expect(erd.length).toBe(3)
  })
})
