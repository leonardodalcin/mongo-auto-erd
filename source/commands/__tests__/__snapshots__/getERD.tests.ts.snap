// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getERD should match snapshot 1`] = `
Array [
  Object {
    "name": "users",
    "properties": Array [
      Object {
        "name": "__v",
        "types": Array [
          "number",
          "number",
          "number",
        ],
        "values": Array [
          0,
          0,
          0,
        ],
      },
      Object {
        "name": "_id",
        "types": Array [
          "objectId",
          "objectId",
          "objectId",
        ],
        "values": Array [
          "5dd059f71bb2d30f42694b57",
          "5dd059f71bb2d30f42694b58",
          "5dd059f71bb2d30f42694b59",
        ],
      },
      Object {
        "name": "name",
        "types": Array [
          "string",
          "string",
        ],
        "values": Array [
          "Test user name",
          "Test user name2",
        ],
      },
    ],
    "relationships": Array [
      Object {
        "propertyNames": Array [
          "_id",
        ],
        "targetCollectionName": "users",
      },
    ],
  },
  Object {
    "name": "photos",
    "properties": Array [
      Object {
        "name": "__v",
        "types": Array [
          "number",
          "number",
        ],
        "values": Array [
          0,
          0,
        ],
      },
      Object {
        "name": "_id",
        "types": Array [
          "objectId",
          "objectId",
        ],
        "values": Array [
          "5dd059f71bb2d30f42694b5a",
          "5dd059f71bb2d30f42694b5b",
        ],
      },
      Object {
        "name": "title",
        "types": Array [
          "string",
          "string",
        ],
        "values": Array [
          "Test photo name",
          "Test photo name2",
        ],
      },
      Object {
        "name": "user",
        "types": Array [
          "objectId",
          "objectId",
        ],
        "values": Array [
          "5dd059f71bb2d30f42694b57",
          "5dd059f71bb2d30f42694b57",
        ],
      },
    ],
    "relationships": Array [
      Object {
        "propertyNames": Array [
          "_id",
        ],
        "targetCollectionName": "photos",
      },
      Object {
        "propertyNames": Array [
          "user",
        ],
        "targetCollectionName": "users",
      },
    ],
  },
  Object {
    "name": "usergroups",
    "properties": Array [
      Object {
        "name": "__v",
        "types": Array [
          "number",
          "number",
        ],
        "values": Array [
          0,
          0,
        ],
      },
      Object {
        "name": "_id",
        "types": Array [
          "objectId",
          "objectId",
        ],
        "values": Array [
          "5dd059f71bb2d30f42694b5c",
          "5dd059f71bb2d30f42694b5d",
        ],
      },
      Object {
        "name": "name",
        "types": Array [
          "string",
          "string",
        ],
        "values": Array [
          "Test user group",
          "Test user group2",
        ],
      },
      Object {
        "name": "users",
        "types": Array [
          "objectId",
          "objectId",
          "objectId",
          "objectId",
        ],
        "values": Array [
          "5dd059f71bb2d30f42694b57",
          "5dd059f71bb2d30f42694b57",
          "5dd059f71bb2d30f42694b57",
          "5dd059f71bb2d30f42694b57",
        ],
      },
    ],
    "relationships": Array [
      Object {
        "propertyNames": Array [
          "_id",
        ],
        "targetCollectionName": "usergroups",
      },
      Object {
        "propertyNames": Array [
          "users",
        ],
        "targetCollectionName": "users",
      },
    ],
  },
]
`;
